partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,757 @@
+use indexmap::IndexMap;
+use partiql_ast::ast::{CaseSensitivity, SymbolPrimitive};
+use partiql_catalog::Catalog;
+use partiql_logical::{BindingsOp, LogicalPlan, OpId, PathComponent, ValueExpr};
+use partiql_types::{
+    any, missing, unknown, ArrayType, BagType, PartiqlType, StructConstraint, StructField,
+    StructType, TypeKind,
+};
+use partiql_value::{BindingsName, Value};
+use petgraph::algo::toposort;
+use petgraph::graph::NodeIndex;
+use petgraph::prelude::StableGraph;
+use std::collections::{BTreeSet, HashMap};
+use thiserror::Error;
+
+#[macro_export]
+macro_rules! ty_ctx {
+    (($x:expr, $y:expr)) => {
+        TypeEnvContext::from(($x, $y))
+    };
+}
+
+#[macro_export]
+macro_rules! ty_env {
+    (($x:expr, $y:expr)) => {
+        LocalTypeEnv::from([($x, $y)])
+    };
+}
+
+const OUTPUT_SCHEMA_KEY: &str = "_output_schema";
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct TypeErr {
+    pub errors: Vec<TypingError>,
+}
+
+#[derive(Error, Debug, Clone, PartialEq, Eq, Hash)]
+#[allow(dead_code)]
+#[non_exhaustive]
+pub enum TypingError {
+    /// Feature has not yet been implemented.
+    #[error("Not yet implemented: {0}")]
+    NotYetImplemented(String),
+    /// Internal error that was not due to user input or API violation.
+    #[error("Illegal State: {0}")]
+    IllegalState(String),
+
+    /// Represents an error in type checking
+    #[error("TypeCheck: {0}")]
+    TypeCheck(String),
+}
+
+#[derive(Debug, Clone)]
+pub enum TypingMode {
+    Permissive,
+    Strict,
+}
+
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+enum LookupOrder {
+    GlobalLocal,
+    LocalGlobal,
+}
+
+/// Represents a type environment context
+#[derive(Debug, Clone)]
+struct TypeEnvContext {
+    env: LocalTypeEnv,
+    /// Represents the type that is used for creating the `env` in the [TypeEnvContext]
+    derived_type: PartiqlType,
```

Comment:
Could you give more explanation for what `derived_type` is used for?

Reply:
Derived type is the type that is used for building the local typing environment in the stack.

Reply:
With the current model this is required esp. to know if we are in an open or closed schema during type-checking. E.g., when requiring to type-check `d` in `a.b.c.d` with `a` being `{'b': {'c': INT}}` where `b` is closed and `c` is open, we need to know somehow have access to the type that `d` _may_ have been resulted from to make a decision about its type i.e. being `any` in this case.